lib/spack/spack/spec.py

```diff
@@ -2551,7 +2564,15 @@ def _mark_concrete(self, value=True):
         Only for internal use -- client code should use "concretize"
         unless there is a need to force a spec to be concrete.
         """
+        # if set to false, clear out all hashes (set to None or remove attr)
+        # may need to change references to respect None
         for s in self.traverse():
+            if (not value) and s.concrete and s.package.installed:
+                continue
+            elif not value:
+                s.clear_cached_hashes()
+            s._normal = value
+            s._concrete = value
```

Comment:
```suggestion
```
These two lines are redundant with the function call below them.

Reply:
Applied in latest commit.